### DISCUSSION TITLE
USB Host to a CP2104 USB to UART Bridge

### DISCUSSION THREAD

**Discussion by wigahluk** [hathach/tinyusb#1831] [Q&A]
Hello,

Apologies in advance if I'm not using the correct terminology, this is my first attempt programming microcontrollers and I'm new to this world. I'm trying to connect a USB device `CP2104` to an RP2040 Pi Pico running as a USB host. My intention is to control the pico with this device. I'm able to run the `bare_api` example and get some data on the Device Descriptor:

```
Device attached, address = 1
Device 1: ID 10c4:ea60
Device Descriptor:
  bLength             18
  bDescriptorType     1
  bcdUSB              0200
  bDeviceClass        0
  bDeviceSubClass     0
  bDeviceProtocol     0
  bMaxPacketSize0     64
  idVendor            0x10c4
  idProduct           0xea60
  bcdDevice           0100
  iManufacturer       1     Silicon Labs
  iProduct            2     CP2104 USB to UART Bridge Controller
  iSerialNumber       3     01D7A115
  bNumConfigurations  1
Device removed, address = 1
```
In a Linux machine, I'm also able to connect with Minicom (`minicom -b 115200 -H -D /dev/ttyUSB0`) to the device. If I use the buttons and physical controls on the device, I see the codes in the terminal.

I haven't been able to read the codes from the device within the pico, I tried the CDC host example (being a USB to UART bridge, I thought this one was my best shot) as well as the HID one (I give it a shot as I was not successful with the CDC case), in both, I see the calls to  `tuh_mount_cb` and `tuh_umount_cb` but none of the specific mount callbacks for CDC nor HID are ever called (like the `tuh_cdc_mount_cb` one).

I also have tried the examples on Arduino (https://github.com/adafruit/Adafruit_TinyUSB_Arduino/blob/master/examples/DualRole/serial_host_bridge/serial_host_bridge.ino) with the same results, I can see when the device is mounted, but cannot read any event from it.

I appreciate any help, thanks!

**Comment by hathach** [comment]
cp210x and ftdi driver isn't supported yet

**Comment by IngHK** [comment]
See #2488.
Any kind of feedback, review and test is welcome.

**Comment by Takatugu-Endou** [comment]
Top url
https://www.hobbytronics.co.uk/
USB Host Controller Board V2.4
https://www.hobbytronics.co.uk/product/host-board
USB Host Board – CDC Class Serial Software
https://www.hobbytronics.co.uk/usb-host-cdc
Test Review(Japanese site)
https://cba.sakura.ne.jp/parts01/kit_i66.htm
-----------------------------------------------------------------
Trying out the PIC USB host function (CDC-Basic)(Japanese site)
http://zattouka.net/GarageHouse/micon/USB/HOST/CDC/CDC.htm
-----------------------------------------------------------------

I think the only solution is to either buy or make it.

**Comment by hathach** [comment]
Updated, it is already supported for some time
